### PR TITLE
Feature/EUI 2878 support combinations of AND and OR boolean logic

### PR DESCRIPTION
**Pull request by RiteshHMCTS**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/EUI-2878

### Change description ###
Evaluation of show hide condition which contains AND and OR conditions

**Does this PR introduce a breaking change?** (check one with "x")

```
[x] Yes
[ ] No
```

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -92,6 +93,7 @@
     "@types/jasmine": "~2.8.0",
     "@types/jasminewd2": "^2.0.3",
     "@types/node": "^10.11.0",
+    "@types/pegjs": "^0.10.2",
```
Comment (paul-graham) [inline package.json:96]: Any reason why this version is not the same as the `pegjs` version above?

File: src/shared/components/case-editor/services/wizard-page-field-to-case-field.mapper.spec.ts [hunk 2]
```diff
@@ -15,7 +15,7 @@ describe('WizardPageFieldToCaseFieldMapper', () => {
 
   let wizardPageFieldToCaseFieldMapper: WizardPageFieldToCaseFieldMapper;
   const INCOMPLETE_SHOW_CONDITION_ONE = 'AddressLine2="test"';
-  const INCOMPLETE_SHOW_CONDITION_TWO = 'AddressLine1CONTAINS"test"';
+  const INCOMPLETE_SHOW_CONDITION_TWO = 'AddressLine1 CONTAINS"test"';
```
Comment (paul-graham) [inline src/shared/components/case-editor/services/wizard-page-field-to-case-field.mapper.spec.ts:18]: What is the reason for this change? I think all operators can be without spaces in the CCD configuration... so if this was to get a test working, it might suggest that there's a problem with the parsing logic. 🤔

File: src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts [hunk 3]
```diff
@@ -350,7 +351,7 @@ describe('conditional-show', () => {
     });
 
     it('field mentioned in condition has no value asked in CONTAINS condition', () => {
-      let sc = new ShowCondition('fieldCONTAINS"test,mest"');
+      let sc = new ShowCondition('field CONTAINS"test,mest"');
```
Comment (paul-graham) [inline src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts:354]: Same as the comment above about operators not needing to have spaces before or after...

File: src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts [hunk 4]
```diff
@@ -360,7 +361,7 @@ describe('conditional-show', () => {
     });
 
     it('field mentioned in condition has no value', () => {
-      let sc = new ShowCondition('fieldCONTAINS"test,mest"');
+      let sc = new ShowCondition('field CONTAINS"test,mest"');
```
Comment (paul-graham) [inline src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts:364]: Same as the comment above about operators not needing to have spaces before or after...

File: src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts [hunk 5]
```diff
@@ -370,7 +371,7 @@ describe('conditional-show', () => {
     });
 
     it('field mentioned in single value condition has no value', () => {
-      let sc = new ShowCondition('fieldCONTAINS"test"');
+      let sc = new ShowCondition('field CONTAINS"test"');
```
Comment (paul-graham) [inline src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts:374]: Same as the comment above about operators not needing to have spaces before or after...

File: src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts [hunk 6]
```diff
@@ -514,7 +515,7 @@ describe('conditional-show', () => {
 
     it('should return true when all conditions are true', () => {
       caseField1.value = ['s1', 's2', 's3'];
-      let sc = new ShowCondition('field1CONTAINS"s3,s2" AND field2=3 AND field3="te*" AND field4="s1 AND s2"');
+      let sc = new ShowCondition('field1 CONTAINS"s3,s2" AND field2=3 AND field3="te*" AND field4="s1 AND s2"');
```
Comment (paul-graham) [inline src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts:518]: Same as the comment above about operators not needing to have spaces before or after...

File: src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts [hunk 7]
```diff
@@ -538,7 +539,7 @@ describe('conditional-show', () => {
 
     it('should evaluate AND conditions correctly for a mix of EQUALS and CONTAINS', () => {
       caseField2.value = ['s4', 's2', 's3'];
-      let sc = new ShowCondition('field4="s1 AND s2" AND field2CONTAINSs3,s4');
+      let sc = new ShowCondition('field4="s1 AND s2" AND field2 CONTAINS "s3,s4"');
```
Comment (paul-graham) [inline src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts:542]: Same as the comment above about operators not needing to have spaces before or after...

File: src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts [hunk 8]
```diff
@@ -548,7 +549,7 @@ describe('conditional-show', () => {
     it('should return true when all conditions are true when using Complex fields', () => {
       caseField1.value = ['s1', 's2', 's3'];
       let sc = new ShowCondition(
-        'field1CONTAINS"s3,s2" AND claimantDetails.AddressUKCode.PostTown="London" AND claimantDetails.AddressUKCode.Country="UK"');
+        'field1 CONTAINS"s3,s2" AND claimantDetails.AddressUKCode.PostTown="London" AND claimantDetails.AddressUKCode.Country="UK"');
```
Comment (paul-graham) [inline src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts:552]: Same as the comment above about operators not needing to have spaces before or after...

File: src/shared/directives/conditional-show/models/peg-result.model.ts [hunk 9]
```diff
@@ -0,0 +1,9 @@
+export class PegCondition {
```
Comment (paul-graham) [inline src/shared/directives/conditional-show/models/peg-result.model.ts:1]: Should be 2 space indent instead of 4.

File: src/shared/directives/conditional-show/services/condition-parser.service.spec.ts [hunk 10]
```diff
@@ -0,0 +1,744 @@
+import { ConditionParser } from './condition-parser.service';
+
+describe('ConditionParser', () => {
+
+    describe('parse', () => {
```
Comment (paul-graham) [inline src/shared/directives/conditional-show/services/condition-parser.service.spec.ts:5]: Indenting should all be 2 spaces rather than 4.

File: src/shared/directives/conditional-show/services/condition-parser.service.ts [hunk 11]
```diff
@@ -0,0 +1,202 @@
+import { _ as _score } from 'underscore';
+import peg from './condition.peg';
+import { FieldsUtils } from '../../../services/fields/fields.utils';
+
+export class ConditionParser {
+
+    private static readonly CONDITION_NOT_EQUALS = '!=';
```
Comment (paul-graham) [inline src/shared/directives/conditional-show/services/condition-parser.service.ts:7]: Indenting should all be 2 spaces rather than 4.

Comment (paul-graham) [inline src/shared/directives/conditional-show/services/condition-parser.service.ts:7]: Maybe make the ones in `ShowCondition` public rather than re-declaring them here?

File: src/shared/directives/conditional-show/services/condition-parser.service.ts [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+                    console.log('Error while parsing number', pathTail[0], e);
+                }
+            }
+        } else {
+            console.log('Path in formArray should start with ', head, ', full path: ', path);
+        }
+    }
+
+    private static removeStarChar(str: string): string {
+        return str.substring(0, str.length - 1);
```
Comment (paul-graham) [inline src/shared/directives/conditional-show/services/condition-parser.service.ts:196]: This felt like slightly risky code when I saw it a while ago and it still feels that way now. It doesn't look at whether or not the parameter even _contains_ an asterisk; it simply removes the last character, whatever that happens to be.

File: src/shared/directives/conditional-show/services/condition-parser.service.ts [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+    }
+
+    private static okIfBothEmpty(right: string, value: any): boolean {
+        return value === null && (right === '');
```
Comment (paul-graham) [inline src/shared/directives/conditional-show/services/condition-parser.service.ts:200]: What about if `value === undefined` or `value === ''` or maybe even `value === []`? 🤔

**Review by paul-graham — CHANGES_REQUESTED** [review]
Mostly a bit concerned about the addition of a space ahead of all `CONTAINS` operators in the unit tests. I think people are able to write it simply as `fieldCONTAINS"bob"` and expect it to work... so if the unit tests have all been changed because that's no longer supported, this will break things.

Also, a lot of the new files have 4 space indenting when it should be 2.

File: src/shared/directives/conditional-show/services/condition-parser.service.ts (hunk 12 above)
Comment (paul-graham) [reply src/shared/directives/conditional-show/services/condition-parser.service.ts:196]: Also, it doesn't do any null checking on `str`.

File: src/shared/components/case-editor/services/wizard-page-field-to-case-field.mapper.spec.ts (hunk 2 above)
Comment (RiteshHMCTS) [reply src/shared/components/case-editor/services/wizard-page-field-to-case-field.mapper.spec.ts:18]: As we have implemented peg parser it doesn't accept CONTAINS operator without space; space should be there between fieldname and CONTAINS operator otherwise it wont parse the condition. This doesn't applicable for =/!= operator.

File: src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts (hunk 3 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts:354]: As I mentioned in my earlier comment

File: src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts (hunk 4 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts:364]: As I mentioned in my earlier comment

File: src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts (hunk 5 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts:374]: As I mentioned in my earlier comment

File: src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts (hunk 6 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts:518]: As I mentioned in my earlier comment

File: src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts (hunk 8 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts:552]: As I mentioned in my earlier comment

File: src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts (hunk 7 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/domain/conditional-show.model.spec.ts:542]: As I mentioned in my earlier comment

File: src/shared/directives/conditional-show/models/peg-result.model.ts (hunk 9 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/models/peg-result.model.ts:1]: Changed it to 2

File: src/shared/components/case-editor/services/wizard-page-field-to-case-field.mapper.spec.ts (hunk 2 above)
Comment (paul-graham) [reply src/shared/components/case-editor/services/wizard-page-field-to-case-field.mapper.spec.ts:18]: The problem is that there are existing configurations out there _without_ a space... so all those configurations will be broken. If the peg parser doesn't work without a preceding space, I'd suggest it needs pre-parsing to globally replace `"CONTAINS"` with `" CONTAINS"`. 🤷

Comment (RiteshHMCTS) [reply src/shared/components/case-editor/services/wizard-page-field-to-case-field.mapper.spec.ts]: as we discussed implemented solution to replace "CONTAINS" with " CONTAINS"

File: src/shared/directives/conditional-show/services/condition-parser.service.spec.ts (hunk 10 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/services/condition-parser.service.spec.ts]: Changed it to 2

File: src/shared/directives/conditional-show/services/condition-parser.service.ts (hunk 11 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/services/condition-parser.service.ts]: changed it to 2

Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/services/condition-parser.service.ts]: as mentioned above made it public in ShowCondition and used it

File: src/shared/directives/conditional-show/services/condition-parser.service.ts (hunk 12 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/services/condition-parser.service.ts]: as mentioned above implemented

File: src/shared/directives/conditional-show/services/condition-parser.service.ts (hunk 13 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/services/condition-parser.service.ts]: this is old code so keeping it as it is.

File: package.json (hunk 1 above)
Comment (paul-graham) [reply package.json:96]: @adnanakgun, are you able to comment on this one?

**Review by paul-graham — COMMENTED** [review]
Thanks for making the requested changes, @RiteshHMCTS. Just that one question remains about the version differences for pegjs...

File: package.json (hunk 1 above)
Comment (adnanakgun) [reply package.json:96]: I am using a tool that imports that stuff automatically. I'll try to match versions and see if that breaks anything. Good catch.

Comment (adnanakgun) [reply package.json]: Made the change @paul-graham .

**Review by paul-graham — DISMISSED** [review]
All concerns are now resolved. Thanks, @RiteshHMCTS and @adnanakgun! 😁 

Approved. 👍

**Comment by rayliangatos** [comment]
> ### JIRA link (if applicable)
> https://tools.hmcts.net/jira/browse/EUI-2878
> 
> ### Change description
> Evaluation of show hide condition which contains AND and OR conditions
> 
> **Does this PR introduce a breaking change?** (check one with "x")
> 
> ```
> [ ] Yes
> [x] No
> ```

I would mark this as breaking changes and it requires full regression tests.

File: src/shared/directives/conditional-show/domain/conditional-show.model.ts [hunk 14]
```diff
@@ -2,42 +2,70 @@ import { _ as _score } from 'underscore';
...
     if (!pathPrefix || pathPrefix === '') {
       return showCondition;
     }
-    if (showCondition.search(ShowCondition.OR_CONDITION_REGEXP) !== -1) {
-      let orConditions = showCondition.split(ShowCondition.OR_CONDITION_REGEXP);
-      orConditions = this.extractConditions(orConditions, pathPrefix);
-      return orConditions.join(' OR ');
+    if (!showCondition) {
+      return showCondition;
+    }
```
Comment (rayliangatos) [inline src/shared/directives/conditional-show/domain/conditional-show.model.ts:23]: this logic is big weird to me, returning something not existing.

File: src/shared/directives/conditional-show/domain/conditional-show.model.ts [hunk 15]
```diff
@@ -2,42 +2,70 @@ import { _ as _score } from 'underscore';
...
+    return processedCondition.replace(/ CONTAINS/g, 'CONTAINS');
+  }
+
+  private static processAddPathPrefixToCondition(formula: any, pathPrefix: string, originalCondition: string): string {
+    let finalCondition: string = originalCondition;
+    if (Array.isArray(formula)) {
+      formula.forEach(condition => {
+        if (typeof condition === 'string' && this.validJoinComparators.indexOf(condition) !== -1) {
+          // do nothing
+        } else {
```
Comment (rayliangatos) [inline src/shared/directives/conditional-show/domain/conditional-show.model.ts:41]: can this if/else be reversed to avoid the '//do nothing'

File: src/shared/directives/conditional-show/domain/conditional-show.model.ts [hunk 16]
```diff
@@ -2,42 +2,70 @@ import { _ as _score } from 'underscore';
```
Comment (rayliangatos) [inline src/shared/directives/conditional-show/domain/conditional-show.model.ts:2]: though it's not in this change but

import { _ as _score } from 'underscore';

is unused import

File: src/shared/directives/conditional-show/domain/conditional-show.model.ts [hunk 17]
```diff
@@ -61,6 +89,27 @@ export class ShowCondition {
     return [ condition.split(separator)[0], separator ];
   }
 
+  private static getConditions(formula: any): string {
+    let conditionList: string[] = [];
+    const newFormula = typeof formula === 'string' ? JSON.parse(formula) : formula;
+    if (!!formula) {
+      if (Array.isArray(newFormula)) {
```
Comment (rayliangatos) [inline src/shared/directives/conditional-show/domain/conditional-show.model.ts:96]: this two ifs can be merged to one to avoid too deep if statements.

File: src/shared/directives/conditional-show/services/condition-parser.service.ts [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+      return expectedValues.every(item => values.search(item) >= 0);
+    } else {
+      let values = currentValue && Array.isArray(currentValue) ? currentValue.toString() : '';
+      return values.search(expectedValue) >= 0;
+    }
+  }
+
+  private static unquoted(str: string): string {
+    const res = str.replace(/^"|"$/g, '');
+    return res;
```
Comment (rayliangatos) [inline src/shared/directives/conditional-show/services/condition-parser.service.ts:148]: const res is redudant it can be returned immediately.

File: src/shared/directives/conditional-show/services/condition-parser.service.ts [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+        const currentValue = this.findValueForComplexCondition(fields, head, tail, path);
+        const expectedValue = this.unquoted(formula.split(conditionSeparator)[1]);
+        if (conditionSeparator === ShowCondition.CONTAINS) {
+          currentConditionResult = this.checkValueContains(expectedValue, currentValue);
+        } else {
+          currentConditionResult = this.checkValueEquals(expectedValue, currentValue, conditionSeparator);
+        }
+      }
+
+      if (isJoinComparator(conditions[index - 1])) {
```
Comment (rayliangatos) [inline src/shared/directives/conditional-show/services/condition-parser.service.ts:57]: can this 4 ifs change to switch/case statements and breakdown to 4 separate methods

File: src/shared/directives/conditional-show/services/condition-parser.service.ts [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+    if (path.startsWith(head)) {
+      const [_, ...pathTail] = path.split(/[_]+/g);
+      if (pathTail.length > 0) {
+        try {
+          let arrayIndex = Number.parseInt(pathTail[0], 10);
+          const [__, ...dropNumberPath] = pathTail;
+          return (fields[head][arrayIndex] !== undefined) ? this.findValueForComplexCondition(
+          fields[head][arrayIndex]['value'], tail[0], tail.slice(1), dropNumberPath.join('_')) : null;
+        } catch (e) {
+          console.log('Error while parsing number', pathTail[0], e);
```
Comment (rayliangatos) [inline src/shared/directives/conditional-show/services/condition-parser.service.ts:186]: should this be console.error

File: src/shared/directives/conditional-show/services/condition-parser.service.ts [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+        }
+      }
+    } else {
+      console.log('Path in formArray should start with ', head, ', full path: ', path);
```
Comment (rayliangatos) [inline src/shared/directives/conditional-show/services/condition-parser.service.ts:190]: from the message it looks like a warning so should it be console.warn or console.error.

File: src/shared/directives/conditional-show/services/condition-parser.service.ts [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+    if (conditionSeparator === ShowCondition.CONDITION_NOT_EQUALS) {
+      return expectedValues !== values;
+    } else {
+      return expectedValues === values;
+    }
+  }
+
+  private static checkValueContains(expectedValue: string, currentValue: any): boolean {
+    if (expectedValue.search(',') > -1) {
+      let expectedValues = expectedValue.split(',').sort();
```
Comment (rayliangatos) [inline src/shared/directives/conditional-show/services/condition-parser.service.ts:137]: is this const or let?

File: src/shared/directives/conditional-show/services/condition-parser.service.ts [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+      let values = currentValue ? currentValue.sort().toString() : '';
```
Comment (rayliangatos) [inline src/shared/directives/conditional-show/services/condition-parser.service.ts:138]: ditto

File: src/shared/directives/conditional-show/services/condition-parser.service.ts [hunk 24]
```diff
@@ -0,0 +1,204 @@
...
+      return expectedValues.every(item => values.search(item) >= 0);
+    } else {
+      let values = currentValue && Array.isArray(currentValue) ? currentValue.toString() : '';
```
Comment (rayliangatos) [inline src/shared/directives/conditional-show/services/condition-parser.service.ts:141]: ditto

File: src/shared/directives/conditional-show/services/condition-parser.service.ts (hunk 20 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/services/condition-parser.service.ts:186]: This is existing code I just moved to this file so want to keep it as is

File: src/shared/directives/conditional-show/services/condition-parser.service.ts (hunk 21 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/services/condition-parser.service.ts:190]: This is existing code.

File: src/shared/directives/conditional-show/domain/conditional-show.model.ts (hunk 16 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/domain/conditional-show.model.ts:2]: removed

File: src/shared/directives/conditional-show/services/condition-parser.service.ts (hunk 18 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/services/condition-parser.service.ts:148]: done

File: src/shared/directives/conditional-show/domain/conditional-show.model.ts (hunk 17 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/domain/conditional-show.model.ts:96]: here in first if I'm checking whether formula field contains any value or not and in next if I'm evaluating the condition after I get value in formula field.
Still I have made small change in this.

File: src/shared/directives/conditional-show/services/condition-parser.service.ts (hunk 22 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/services/condition-parser.service.ts:137]: changed it to const

File: src/shared/directives/conditional-show/services/condition-parser.service.ts (hunk 24 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/services/condition-parser.service.ts:141]: changed it to const

File: src/shared/directives/conditional-show/services/condition-parser.service.ts (hunk 23 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/services/condition-parser.service.ts:138]: changed it to const

File: src/shared/directives/conditional-show/domain/conditional-show.model.ts (hunk 14 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/domain/conditional-show.model.ts:23]: changed the logic

File: src/shared/directives/conditional-show/domain/conditional-show.model.ts (hunk 15 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/domain/conditional-show.model.ts:41]: as condition contains different type of value and I want to skip it when its type is string.

File: src/shared/directives/conditional-show/services/condition-parser.service.ts (hunk 19 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/services/condition-parser.service.ts:57]: it's bit complicated; but what I did is moved some code to new function.

File: src/shared/directives/conditional-show/services/condition-parser.service.ts (hunk 21 above)
Comment (rayliangatos) [reply src/shared/directives/conditional-show/services/condition-parser.service.ts:194]: if this is copied from exsiting code, please also change it.

File: src/shared/directives/conditional-show/services/condition-parser.service.ts (hunk 20 above)
Comment (rayliangatos) [reply src/shared/directives/conditional-show/services/condition-parser.service.ts:190]: if this is copied from exsiting code, please also change it.

Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/services/condition-parser.service.ts]: changed to console.error

File: src/shared/directives/conditional-show/services/condition-parser.service.ts (hunk 21 above)
Comment (RiteshHMCTS) [reply src/shared/directives/conditional-show/services/condition-parser.service.ts]: changed to console.error

**Review by rayliangatos — APPROVED** [review]
Approved, QA to be cautious it needs full regression tests.

**Comment by phillip-whitaker-hmcts** [comment]
reopening branch as part of https://tools.hmcts.net/jira/browse/EUI-3729

**Review by tarikdogan-hmcts — DISMISSED** [review]

**Review by tarikdogan-hmcts — DISMISSED** [review]

**Review by Mo-Lala84 — DISMISSED** [review]
approved
